### PR TITLE
[perf] SyncBatchNorm: avoid 2nd set of all_reduce when wrapped by checkpoint_wrapper

### PR DESCRIPTION
**Pull request by msbaines**
This change also ensure that we calculate running_{mean,var} correctly
when wrapped.

## What does this PR do?
Fixes # (issue).

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: fairscale/experimental/nn/sync_batchnorm.py [hunk 1]
```diff
@@ -138,22 +118,27 @@ def __init__(
...
         super().__init__(*args, **kwargs)  # type: ignore
         self._process_group = process_group if process_group is not None else dist.group.WORLD
+        self.saved_for_2nd_fwd: Optional[Tuple] = None
 
-    def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
+    def forward(self, input: Tensor) -> Tensor:  # type: ignore
         if not dist.is_initialized() or not self.training:
             return super().forward(input)
 
+        if not hasattr(self, "_checkpoint_fwd_counter") or self._checkpoint_fwd_counter == 0:
```
Comment (min-xu-ai) [inline fairscale/experimental/nn/sync_batchnorm.py:127]: The `_checkpoint_fwd_counter` currently is used by FSDP for the case where a module is forwarded multiple times in a single outer forward pass. In SyncBN's case, do we actually only want to keep the stats once in that case?

i.e.:

```
class Layer():
    self.bn = sync_bn()
    def forward():
       ...
       x = self.bn(x)
       ...

class Mod():
       self.layer = Layer()
       def forward():
             return layer(layer(x))
```

In the above case, the bn inside layer is called 4 times if `Mod` is wrapped by a checkpoint. In torch BN's case, the running stats will be kept in 2 of the 4 times. In this change, using `_checkpoint_fwd_counter`, however, it will be kept in only 1 out of 4 times. I think 2 is probably the right thing to do.

Maybe, this BN should keep a `track_running_stats` flag like torch's BN? That way, checkpoint_wrapper will just use `patch_batchnorm` to toggle that flag to get the right behavior?

File: tests/experimental/nn/test_sync_batchnorm.py [hunk 2]
```diff
@@ -92,6 +93,19 @@ def parity3d_bn():
...
+def parity3d_checkpoint_syncbn():
+    rank = dist.get_rank()
+    torch.cuda.set_device(rank)
+    torch.manual_seed(rank)
+
+    x = torch.randn(4, 3, 4, 4, 4).cuda() * rank
+    torch_bn = torch.nn.SyncBatchNorm(3).cuda()
+    fs_bn = SyncBatchNorm(3).cuda()
+    fs_bn = checkpoint_wrapper(fs_bn, maintain_forward_counter=True)
+    check_parity_ddp(torch_bn, fs_bn, x)
```
Comment (min-xu-ai) [inline tests/experimental/nn/test_sync_batchnorm.py:106]: does check parity function actually check the buffer values? I think the running stats are kept in the buffers and they are not used during training and only later get used during inference? Sorry if I misunderstood that part.

File: fairscale/experimental/nn/sync_batchnorm.py (hunk 1 above)
Comment (msbaines) [reply fairscale/experimental/nn/sync_batchnorm.py:127]: This BN does support track_running_stats. I verified parity is correct for a single forward pass. Did not think about 2 forward passes. Will add a test and code for that.

File: tests/experimental/nn/test_sync_batchnorm.py (hunk 2 above)
Comment (msbaines) [reply tests/experimental/nn/test_sync_batchnorm.py:106]: check_parity_ddp checks that the buffer values match:

https://github.com/facebookresearch/fairscale/blob/d240b748546445374e3823b4f8c3557963ae77f5/tests/experimental/nn/test_sync_batchnorm.py#L74

**Comment by msbaines** [comment]
Added support for calling SyncBatchNorm twice in a checkpoint_wrapper.

File: fairscale/nn/checkpoint/checkpoint_activations.py [hunk 3]
```diff
@@ -18,6 +19,71 @@
...
+
+            def forward(self, input):
+                if not is_recomputing():
+                    self.counter += 1
+                return input
+
+    Returns:
+        bool: :data:`True` if it's under checkpoint recomputation.
+
+    .. see also:: :ref:`Detecting Recomputation`
```
Comment (min-xu-ai) [inline fairscale/nn/checkpoint/checkpoint_activations.py:81]: where is the ref defined? I am not sure it is supposed to link to somewhere else or not.

**Review by min-xu-ai — APPROVED** [review]
This is very cool! Some nonblocking suggestions.

File: fairscale/nn/checkpoint/checkpoint_activations.py [hunk 4]
```diff
@@ -18,6 +19,71 @@
 from .checkpoint_utils import dec_counter, inc_counter, init_counter, patch_batchnorm
 
 
+class ThreadLocal(threading.local):
+    def __init__(self) -> None:
+        self.is_checkpointing = False
+        self.is_recomputing = False
+
+
+thread_local = ThreadLocal()
```
[the remark this answers is not in the document]
Comment (msbaines) [reply fairscale/nn/checkpoint/checkpoint_activations.py:30]: A global would be not be thread-safe. I believe PyTorch also uses a thread-local for this sort of thing.

File: fairscale/experimental/nn/sync_batchnorm.py [hunk 5]
```diff
@@ -138,22 +120,32 @@ def __init__(
     ) -> None:
         super().__init__(*args, **kwargs)  # type: ignore
         self._process_group = process_group if process_group is not None else dist.group.WORLD
+        self.saved_for_2nd_fwd: List[Tuple] = []
+        self.disable_patch_batchnorm = True
```
Comment (anj-s) [inline fairscale/experimental/nn/sync_batchnorm.py:124]: can we add a comment about what this property is about?

File: tests/experimental/nn/test_sync_batchnorm.py [hunk 6]
```diff
@@ -92,6 +97,34 @@ def parity3d_bn():
...
+
+
+@pg_test()
+def parity3d_checkpoint_syncbn_twice():
```
Comment (anj-s) [inline tests/experimental/nn/test_sync_batchnorm.py:114]: how does this test 2 forward passes?

Comment (msbaines) [reply tests/experimental/nn/test_sync_batchnorm.py:114]: This tests uses an nn.Sequential to call the same BatchNorm instance twice.

File: fairscale/nn/checkpoint/checkpoint_activations.py [hunk 7]
```diff
@@ -18,6 +19,70 @@
...
+    orig = thread_local.is_checkpointing
+    thread_local.is_checkpointing = True
+    try:
+        yield
+    finally:
+        thread_local.is_checkpointing = orig
+
+
+@contextmanager
+def enable_recomputing() -> Generator[None, None, None]:
```
Comment (anj-s) [inline fairscale/nn/checkpoint/checkpoint_activations.py:45]: nit: we should ideally test this in isolation of SyncBN. WDYT?

File: tests/experimental/nn/test_sync_batchnorm.py (hunk 6 above)
Comment (anj-s) [reply tests/experimental/nn/test_sync_batchnorm.py:114]: ah got it! it's the same instance. For some reason I was thinking of two explicit forward calls and checking stats in between.

File: fairscale/nn/checkpoint/checkpoint_utils.py [hunk 8]
```diff
@@ -42,7 +42,7 @@ def post_forward(module: _BatchNorm, input: Tensor, result: Tensor) -> None:
     hooks = []
     for name, child in module.named_modules():
         # _BatchNorm is base for bn1d, bn2d, bn3d and sync_bn, apex_sync_bn, etc.
-        if isinstance(child, _BatchNorm):
+        if isinstance(child, _BatchNorm) and not hasattr(child, "disable_patch_batchnorm"):
```
Comment (anj-s) [inline fairscale/nn/checkpoint/checkpoint_utils.py:45]: again, a comment about this property would help here as well.

**Comment by anj-s** [comment]
@msbaines  this is really nice and simple to understand :) Sorry for the late review. Feel free to address (or ignore if not required) in any later PRs you have in the pipeline.
